### ISSUE TITLE
Connection seems dead for 30s, then closes and starts a new one.

### ISSUE THREAD

**Issue by twdkeule** [lucas-clemente/quic-go#624]
I have an issue where a QUIC video stream seems dead for about 30 seconds before restarting a new connection (new CHLO etc.), this happens about 8 times in the 10min video.

Setup:
- Server: Caddy using QUIC
- Client: Chromium 58.0.3029.96 with force-quic-on server
- node in between to limit throughput with tcconfig (using netem/tc) and tcpdump capture

Might be related to #576 because:

1. The last packet sent by the server is at 03:31:24.277
2. everything is congestion limited
3. The client keeps sending Ack frames (I suppose for earlier video segments)
4. Everythin seems normal up until (and including) 03:31:35.415
5. Then no traffic for 20 seconds after which the client sends something strange:
```
frames.RstStreamFrame{StreamID:0x7, ErrorCode:0x6, ByteOffset:0x0}
Ignoring error in session: RST_STREAM received for unknown stream
```
6. This is ignored, but another 20 seconds later, the client closes the connection: `ConnectionCloseFrame{ErrorCode:0x19, ReasonPhrase:"No recent network activity."}`\
7. After this the client starts a new connection an everything is fine for a while.

I have a Wireshark capture, but it is 140MB because of the video. 
 [edit: removed log as it cluttered the conversation]

**Comment by lucas-clemente** [comment]
Thanks a lot for the report! This looks like a duplicate of #555, but I'll keep it around since your log is super helpful :)

In particular, this looks interesting:

```
Congestion limited: bytes in flight 65170, window 64240
[...]
Queueing packet 0x131 for retransmission (RTO), 49 outstanding
Queueing packet 0x132 for retransmission (RTO), 48 outstanding
Congestion limited: bytes in flight 62968, window 2920
```

The congestion window collapses much faster than I would expect here, there may be a bug in the congestion controller.

**Comment by twdkeule** [comment]
Here is the whole log: https://www.dropbox.com/s/kjmwtgh3qqrq4uw/quic.txt?dl=0 (25MB)
It happens 8 times in the whole log, just look for the CHLOs and work back from there.

I wanted to add the Wireshark capture, but then I remembered it can't decrypt the QUIC traffic anyway. Do you know of any way to decrypt the traffic?

**Comment by lucas-clemente** [comment]
Thanks! Don't worry about the wireshark capture, our log should contain all the information we need (and way more than the capture). There's no way to decrypt it unless you somehow export the keys used from Chrome or our implementation, which I don't think is possible.

**Comment by twdkeule** [comment]
Hi,
I actually need this fixed to continue working, so I would like to help fixing it.
From what I read in the other issues the main problem is: the server needs to keep sending control frames (ACKs) when congestion limited. But for that the packetpacker and ackhandler need to be updated to handle non-retransmissible packets.
Can you point me in the right direction to (help) fix(ing) this?
Or do you have any timetable yourselves for fixing this?

**Comment by lucas-clemente** [comment]
I think the ACK retransmittability issue is orthogonal to this one and won't actually resolve it. This is due to how RTOs currently work, I've explained this a bit in https://github.com/lucas-clemente/quic-go/issues/555#issuecomment-298089220.

Tbh I think it's easiest if I fix this myself, but I need to talk to some transport people to figure out the right way. I will try to reach out this week and get a fix together early next week, but I'm quite loaded right now so I can't really promise anything.

**Comment by twdkeule** [comment]
So when the ACKs are being sent, the congestion control issue is still there, but now the client won't disconnect because it can still sent PINGs?

**Comment by lucas-clemente** [comment]
Yep. The fix will probably either involve less exponentiality in the RTOs (e.g. only once per RTT, etc), or faster decrease of bytes in flight.

**Comment by twdkeule** [comment]
I don't think this is the cause, but Google (proto-quic) limits the exponential back-off to [10 RTOs](https://github.com/google/proto-quic/blob/master/src/net/quic/core/quic_sent_packet_manager.cc#L835) (so *1024). 
This would result in a different RTO timeout if the RTT is smaller than 60ms (which it was for this issue).

**Comment by lucas-clemente** [comment]
FWIW this max is dropped in the current loss recovery draft: https://quicwg.github.io/base-drafts/draft-ietf-quic-recovery.html#pseudocode We still have the 60ms limit though: https://github.com/lucas-clemente/quic-go/blob/master/ackhandler/sent_packet_handler.go#L376 I opened https://github.com/quicwg/base-drafts/issues/556.

Again, I'll look into this asap.

**Comment by lucas-clemente** [comment]
I just merged a (temporary) workaround for this in #637. Would you mind re-trying your test, and, if you still see the issue, attach another log? :)

**Comment by twdkeule** [comment]
It seems to be working. I still get the window collapse to 2 packets from the RTO but now there continues to be traffic.
I will do some more tests

```
2017/06/06 15:24:17 [QUIC-GO] Congestion limited: bytes in flight 44981, window 43800
2017/06/06 15:24:17 [QUIC-GO] 	Queueing packet 0xd5cf for retransmission (RTO), 34 outstanding
2017/06/06 15:24:17 [QUIC-GO] 	Queueing packet 0xd5d0 for retransmission (RTO), 33 outstanding
2017/06/06 15:24:17 [QUIC-GO] Congestion limited: bytes in flight 42281, window 2920
2017/06/06 15:24:17 [QUIC-GO] 	Dequeueing retransmission for packet 0xd5d0
2017/06/06 15:24:17 [QUIC-GO] 	Dequeueing retransmission for packet 0xd5cf
2017/06/06 15:24:17 [QUIC-GO] -> Sending packet 0xd5f1 (1350 bytes), forward-secure
2017/06/06 15:24:17 [QUIC-GO] 	-> &frames.StopWaitingFrame{LeastUnacked: 0xd5d1, PacketNumberLen: 0x2}
2017/06/06 15:24:17 [QUIC-GO] 	-> &frames.StreamFrame{StreamID: 313, FinBit: false, Offset: 0xe2ac, Data length: 0x52f, Offset + Data length: 0xe7db}
2017/06/06 15:24:17 [QUIC-GO] Congestion limited: bytes in flight 43631, window 2920

**Comment by twdkeule** [comment]
I streamed a video a few more times but didn't get any connection issues anymore.
At first sight, the performance is not worse, but I didn't measure it

**Comment by twdkeule** [comment]
It seems the performance is still quite bad for higher RTT values.
![link_utilization_rtt](https://user-images.githubusercontent.com/11250711/27178865-36c4c178-51cc-11e7-8f16-7ff90a5c1917.png)
I will upload a log

**Comment by lucas-clemente** [comment]
Wow, that looks super cool!

We haven't made any performance comparisons to TCP so far, so this is somewhat expected. I would think it's either a bug in the congestion controller or the ackhandler.

I'll have a look at the log (thanks!), but I'd also like to have a reproducible test setup for this case. Probably the easiest would be to add a new benchmark test with simulated higher RTT, or a Chrome integration test with our UDP proxy.

**Comment by lucas-clemente** [comment]
Looking at the first drop in the log, it looks like Chrome has a slight delay in acking some packets, and we trigger an RTO. As we don't have spurious RTO detection yet, this causes the throughput to drop. We should implement that.

I don't yet know why we never recover from that, looking into that now.

**Comment by lucas-clemente** [comment]
Hm, it seems to always happen at the end of a segment download. That would also explain why it initially doesn't happen, as you're downloading many segments at the same time.

**Comment by lucas-clemente** [comment]
I think I've got the issue – we currently RTO based on non-retransmittable packets, which Chrome (correctly) doesn't immediately ACK. The issue for that is #505, I'll start work on that asap.

**Comment by lucas-clemente** [comment]
So, thanks again for bringing this to our attention! :)

One minor point, for future logs it would be super helpful if you set the Go logger to output finer timestamps. You can do that using `log.SetFlags(log.Ldate | log.Ltime | log.Lmicroseconds)`, see https://golang.org/pkg/log/#pkg-constants.

**Comment by twdkeule** [comment]
Actually this plot shows the relation to the downloading of segmetns better:
![buffer_plot](https://user-images.githubusercontent.com/11250711/27183593-637663a8-51df-11e7-8262-c68c7180b7b3.png)
As soon as segments are not downloaded at the same time, the issue starts.

I am making a log with millisecond timestamps :)

**Comment by twdkeule** [comment]
Here is a log woth milliseconds markings: https://www.dropbox.com/s/9mfuy5xcxpcio7f/quic.log?dl=0

**Comment by lucas-clemente** [comment]
Thanks! I've started working on fixing #505, I'll let you know when I have a branch for you to try out :)

**Comment by lucas-clemente** [comment]
If you have time, you can try the fix-505 branch (PR #688), I fixed 3 issues there that might have affected your test.

**Comment by twdkeule** [comment]
It seems performance is greatly improved. I ran tests on commit 1aa9b68a349fe1d17717298517135e980d47c07b.
![link_utilization](https://user-images.githubusercontent.com/11250711/27373817-fdb8b6e4-5669-11e7-9874-d4add2e6a1ff.png)
You can see there is still an issue when the RTT is 100ms. There are a few moments when the perceived throughput drops significantly.
I have the log of that session if you are interested: https://www.dropbox.com/sh/eztae9j38fp0c8j/AAD62Qm45XK-ExZQCuIpMddQa?dl=0

**Comment by marten-seemann** [comment]
Wow, that's amazing. I'm really glad to see that our performance is now comparable to TCP. Thanks for sharing these results!

**Comment by lucas-clemente** [comment]
It really looks very nice, thanks for doing this! I think we should be able to exceed TCP here (especially for link utilization), but that's gonna be much harder to debug…

I'll look into the log soon. If you have more data in the future (e.g. with packet loss), please feel free to share, I'm sure there are more bugs hiding there.

**Comment by twdkeule** [comment]
I guess this can be closed?

**Comment by marten-seemann** [comment]
@twdkeule: I haven't seen this issue in quite a while, so we can close this now.
Please feel free to open new issues if you encounter any other problems.
